libcxx/include/__algorithm/ranges_find_last_if.h

```diff
@@ -0,0 +1,104 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___ALGORITHM_RANGES_FIND_LAST_IF_H
+#define _LIBCPP___ALGORITHM_RANGES_FIND_LAST_IF_H
+
+#include <__concepts/assignable.h>
+#include <__config>
+#include <__functional/identity.h>
+#include <__functional/invoke.h>
+#include <__iterator/concepts.h>
+#include <__iterator/next.h>
+#include <__iterator/projected.h>
+#include <__ranges/access.h>
+#include <__ranges/concepts.h>
+#include <__ranges/dangling.h>
+#include <__ranges/range_adaptor.h>
+#include <__ranges/subrange.h>
+#include <__utility/move.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_STD_VER >= 23
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+namespace ranges {
+
+template <class _Ip, class _Sp, class _Pred, class _Proj>
+_LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr subrange<_Ip>
+__find_last_if_impl(_Ip __first, _Sp __last, _Pred&& __pred, _Proj&& __proj) {
+  if constexpr (common_range<_Ip> && bidirectional_range<_Ip>) {
+    // Optimized for common_range and bidirectional_range
+    _Ip __original_last = __last; // Save the original value of __last
+    while (__first != __last) {
+      --__last;
+      if (std::invoke(__pred, std::invoke(__proj, *__last))) {
+        return {__last, __original_last}; // Directly return the result
+      }
+    }
+    return {__last, __original_last};
+  } else if constexpr (bidirectional_iterator<_Ip> && assignable_from<_Ip&, _Sp&>) {
+    // For non-common but bidirectional ranges
+    _Ip __result        = ranges::next(__first, __last);
+    _Ip __original_last = __result;
+    while (__first != __result) {
+      --__result;
+      if (std::invoke(__pred, std::invoke(__proj, *__result))) {
+        return {__result, __original_last};
+      }
+    }
+    return {__result, __original_last};
+  } else {
+    // Code for the non-bidirectional case
+    _Ip __original_first = __first;
+    _Ip __result         = __first;
+    while (__first != __last) {
+      if (std::invoke(__pred, std::invoke(__proj, *__first))) {
+        __result = __first;
+      }
+      ++__first;
+    }
+    return {__result, __original_first};
```

Comment:
The return value is not correct. you want to return the incremented iterator which equals to last
> Returns: Let i be the last iterator in the range [first, last) for which E is true[.](https://eel.is/c++draft/alg.find.last#2.sentence-1) Returns {i, last}, or {last, last} if no such iterator is found[.](https://eel.is/c++draft/alg.find.last#2.sentence-2)

Please add tests after the fix